Network wide limits (#1362)
vegaprotocol/specs: protocol/0078-NWLI-network_wide_limits.md

### Change of network parameter

If the network parameter gets increased via a governance vote no further actions are needed.\
If it gets decreased below the current total market count then no further actions are needed. Specifically: all existing markets continue functioning in their respective states, but no further proposals are accepted until the count drops below the new limit.

## Parties in a market

Introduce a new network parameter `limits.markets.maxParties` controlling the maximum number of parties allowed within any given market. Default value: `100,000`.

A party gets counted towards the limit if it has either open orders or open positions in the market. Once party has no open orders and no open positions it gets removed from a **total party count** within the market. Once the limit gets reached the market accepts no further orders from parties that are not already in the market.
Once the total party count drops below the limit market accepts orders from any party again (provided its [state](./0043-MKTL-market_lifecycle.md#market-lifecycle-statuses) allows that).

> The limit does not apply to liquidity providers.

Reviewer: I'm not sure I understand? Does that mean that a party which is a liquidity order doesn't take part of this limit? Or that if this limit is reached, and an LP want to join AND the max LP limit is not reached, then the new lp submission should be accepted?

### Change of network parameter

If the network parameter gets increased via a governance vote no further actions are needed.\

> If it gets decreased below the current total party count in any given market then no further actions are needed. Specifically: all parties in that market remain unaffected (their positions and orders remain open), but no new parties are allowed to participate in the market until the total party count drops below the new limit.

Reviewer: Would it make sense to say -> all parties with open orders and no open positions are removed? I'm not sure, and it'd complicate a little bit things though

## Limit orders on a book

> Introduce a new network parameter `limits.markets.maxLimitOrders` controlling the maximum number of limit orders that can rest on a book in any given market.\

Reviewer: Would it make sense to have that per side instead of per book? Just to make it less likely to have a book too massively filled on a specific side? that might also be an acceptable behaviour/.

* if the order is aggressive (results in a trade) the trade proceeds as normal,
* if the order or its part is passive (it doesn't match in full on entry to the matching engine) and it's part would get added to the order book (it's a persistent order) that part of the order gets rejected.

Pegged orders and [liquidity provision orders](./0038-OLIQ-liquidity_provision_order_type.md) do not get counted towards the limit.

### Change of network parameter

If the network parameter gets increased via a governance vote no further actions are needed.\
If it gets decreased below the current total number of orders on the book in any given market then no further actions are needed. Specifically: all open orders in that market remain unaffected, but no new orders are allowed until the **total open order count** drops below the new limit.

## Pegged orders on a market

> Introduce a new network parameter `limits.markets.maxPeggedOrders` controlling the maximum number of pegged orders that can be active in any given market.\

Reviewer: Shouldn't that be a limit per party? it is so low that it seems that a few parties could easily reach that point?

> Default value: `10,000`.

Reviewer 2: ```suggestion
Default value: `1,500`.
```

Reviewer 2: https://vegaprotocol.slack.com/archives/C03GJPCERFY/p1666969465173389

Reviewer 2: Frome @peterbarrow 's testing. 

![image](https://user-images.githubusercontent.com/62546419/198696770-1938cc00-9832-46e5-b12c-a793bb164a27.png)

### Change of network parameter

If the network parameter gets increased via a governance vote no further actions are needed.\
If it gets decreased below the current total number of orders on the book in any given market then no further actions are needed. Specifically: all pegged orders already present in that market remain unaffected, but no new pegged orders are allowed until the **total pegged order count** drops below the new limit.

[Liquidity provision orders](./0038-OLIQ-liquidity_provision_order_type.md) do not get counted towards the limit.

## LP order shapes

> Each [LP order shape](./0038-OLIQ-liquidity_provision_order_type.md#how-they-are-submitted) has a hardcoded limit of at most 20 entries (offsets).

Reviewer: This seems to go against the current default of 100 and maximum per side of 1000. Althought I'd rather reduce it a lot as suggested

## LPs in a market

Introduce a new network parameter `limits.markets.maxLPs` controlling the maximum number of liquidity providers that can be active in any market.\

> Default value: `1000`.

Reviewer 2: I suspect this is too high because it is somehow potentially equivalent to `1 000 x 20 x 2 = 40k` pegged orders but let's wait for @peterbarrow 's comment. Even if we don't resolve this my suggestion would be to pick a default of `50` so that the PR can be merged.

Author: Ok, changed to 50 now.

Each liquidity provider that successfully submits a liquidity provision transaction for a given market gets counted towards the **active LP count** for that market. When an LP in a given market reduces their commitment amount to 0 or gets closed out the LP count for that market gets decremented by 1.

> Once the active LP count in a given market reaches the limit only the LP transactions with commitment amount larger then the lowest commitment already active in that market are accepted. If such a situation occurs the existing LP commitment with the lowest value which has been in the market the shortest should get forcibly cancelled.\

Reviewer 2: I am not sure we want that. Imagine I was the first LP on the market; small commitment but *huge* ELS. This would be a way for someone to force me out.

Author: Good point, committed your suggestion.

> Once the active LP count in a given market reaches the limit only the LP transactions with commitment amount larger then the lowest commitment already active in that market are accepted. If such a situation occurs the existing LP commitment with the lowest value which has been in the market the shortest should get forcibly cancelled.\

Reviewer 2: ```suggestion
Once the active LP count in a given market reaches the limit only the LP transactions with commitment amount implying [equity-like-share](0042-LIQF-setting_fees_and_rewarding_lps.md) larger then the lowest equity-like-share already active in that market are accepted. If such a situation occurs the existing LP commitment with equity-like-share which is getting replaced have their commitment cancelled (bond returned to margin account, volume no longer placed on book but positions and orders kept in place).\
```

* When the limit is reached in the market the [market data](./0021-MDAT-market_data_spec.md) API correctly indicates that. (<a name="0078-NWLI-023" href="#0078-NWLI-023">0078-NWLI-023</a>)
* When the `limits.markets.maxPeggedOrders` limit is reached in the market, but `limits.markets.maxLPs` is not active yet it is still possible to successfully submit a new liquidity provision order (<a name="0078-NWLI-028" href="#0078-NWLI-028">0078-NWLI-028</a>).

### LP order shapes

* Submitting a [liquidity provision order](./0038-OLIQ-liquidity_provision_order_type.md) with 20 buy shapes and 20 sell shapes proceeds without any errors and liquidity provision becomes active. (<a name="0078-NWLI-014" href="#0078-NWLI-014">0078-NWLI-014</a>)
* Submitting a liquidity provision order with 21 buy shapes and 20 sell shapes results in a failure. Error message attributes the rejection to the limit. (<a name="0078-NWLI-015" href="#0078-NWLI-015">0078-NWLI-015</a>)
* Submitting a liquidity provision order with 20 buy shapes and 21 sell shapes results in a failure. Error message attributes the rejection to the limit. (<a name="0078-NWLI-016" href="#0078-NWLI-016">0078-NWLI-016</a>)

### LPs

> * Attempt to place an additional liquidity provision with valid commitment amount below the lowest one in the market when the number of liquidity provisions in the market is equal to `limits.markets.maxLPs` results in a rejection. Error message attributes the rejection to the limit. (<a name="0078-NWLI-017" href="#0078-NWLI-017">0078-NWLI-017</a>)

Reviewer 2: This will potentially need updates, depending on how the discussion above pans out.

Author: Committed your ELS suggestion and modified the ACs.

# Network wide limits

This spec describes network wide limits aimed at keeping the overall system performant and responsive.
Vega has been designed with low-latency in mind so that the responsiveness of the network doesn't unduly impact the trading experience. Furthermore, the current implementation relies on an interplay between the execution layer (`core`) and the data consumption layer (`data node`). For these reasons limits allowing to exercise a degree of control over the computational and data generation loads need to be put in place. These limits need to be adjustable so that each network/shard can be setup for maximum accessibility (hardware constraints permitting) while preserving the desirable user experience and system stability, and mitigating the possibility of a malicious actor trying to deliberately disrupt the system by flooding it with superfluous requests.

## Number of assets

`limits.assets.max`

> TODO: Do we really care? Each asset is added via a governance vote, so it doesn't seem likely we'll get a ton of unnecessary ones (plus we'd then have to implement a process of removing an asset with a vote if the limit was already reached, but there was some new asset that community perceived as very important). Also, does an additional asset itself really increase the load that much?

Reviewer 2: I agree - we shouldn't have a limit on assets unless we have a process for removing them (which is super unclear TBH).

Author: Deleted now.

## Number of markets

> Introduce a new network parameter `limits.markets.max` controlling the maximum number of markets allowed within a network.

Author: In the future we may want to make these market parameters, so some markets that are deemed to be important and constraint by the limits are allowed to grow without increasing the overall limit, but making them network parameters is a sensible first implementation.

> Each market in a `Proposed` [state](./0043-MKTL-market_lifecycle.md) should increment the **total market count** by 1. If a market reaches a `Rejected`, `Cancelled`, `Closed` or `Settled` state the total market count should be decremented by 1.

Reviewer 2: I would only increment once in opening auction. Otherwise someone with enough tokens to submit a makret proposal can trivially exhaust the max.

Reviewer 2: If the market is approved by voting *but* it would exceed the max then it's the same outcome as if it was rejected by voting.

Author: Good point, modified now.

As soon as the total market count reaches the value specified by `limits.markets.max` parmeter no other further market proposals should be accepted. If the total market count drops below `limits.markets.max` the network should start accepting market proposals again.

### Change of network parameter

> If the network parameter gets increased via a governance vote no further actions are needed.\

Author: This is just a line drop for readability.

Reviewer 2: 👍

If it gets decreased below the current total market count then no further actions are needed. Specifically: all existing markets continue functioning in their respective states, but no further proposals are accepted until the count drops below the new limit.

## Parties in a market

Introduce a new network parameter `limits.markets.maxParties` controlling the maximum number of parties allowed within any given market.

> A party gets counted towards the limit if it has either open orders or open positions in the market. Once party has no open orders and no open positions it gets removed from a **total party count** within the market. Once the limit gets reached the market accepts no further orders.

Reviewer 2: ```suggestion
A party gets counted towards the limit if it has either open orders or open positions in the market. Once party has no open orders and no open positions it gets removed from a **total party count** within the market. Once the limit gets reached the market accepts no further orders from parties that are not already in the market.
```

> Once the total party count drops below the limit market accepts orders again (provided its [state](./0043-MKTL-market_lifecycle.md) allows that).

Reviewer 2: ```suggestion
Once the total party count drops below the limit market accepts orders from any party again (provided its [state](./0043-MKTL-market_lifecycle.md) allows that).
```

The limit does not apply to liquidity providers.

### Change of network parameter

> If the network parameter gets increased via a governance vote no further actions are needed.\

Reviewer 2: ```suggestion
If the network parameter gets increased via a governance vote no further actions are needed.
```

If it gets decreased below the current total party count in any given market then no further actions are needed. Specifically: all parties in that market remain unaffected (their positions and orders remain open), but no new parties are allowed to participate in the market until the total party count drops below the new limit.

## Limit orders on a book

> Introduce a new network parameter `limits.markets.maxLimitOrders` controlling the maximum number of limit orders that can rest on a book in any given market. If the limit is reached limit orders can still be submitted, but:

Reviewer 2: A limit order of arbitrary volume contributes +1 to the count.

Author: Added now.

* if the order is aggressive (results in a trade) the trade proceeds as normal,
* if the order is passive (it doesn't match on entry) and would get added to the order book it gets rejected.

> Pegged orders and [liquidity provision orders](./0038-OLIQ-liquidity_provision_order_type.md) do not get counted towards the limit.

Author: The consequence of that is that if the maximum number of orders gets reached in a market then LPs can only exercise their strategy via the automated order shapes. This may not be ideal, but allowing exclusions for LPs would complicate the design and also open up the door for malicious party entering the market as an LP to circumvent the limits, so I'd leave it as is specified here for the first implementation and change it only if we can clear requests from LPs that it impedes their ability to exercise their strategy.

> * if the order is passive (it doesn't match on entry) and would get added to the order boo (it's a persistent order) it gets rejected.

Reviewer 2: ```suggestion
* if the order is passive (it doesn't match on entry) and would get added to the order book (it's a persistent order) it gets rejected.
```

### Change of network parameter

If the network parameter gets increased via a governance vote no further actions are needed.
If it gets decreased below the current total number of orders on the book in any given market then all pegged orders already present in that market remain unaffected, but no new pegged orders are allowed until the **total pegged order count** drops below the new limit.

[Liquidity provision orders](./0038-OLIQ-liquidity_provision_order_type.md) do not get counted towards the limit.

## LP order shapes

Each [LP order shape](./0038-OLIQ-liquidity_provision_order_type.md#how-they-are-submitted) has a hardcoded limit of at most 20 entries (offsets).

## LPs in a market

Introduce a new network parameter `limits.markets.maxLPs` controlling the maximum number of liquidity providers that can be active in any market. Each liquidity provider that successfully submits a liquidity provision transaction for a given market gets counted towards the **active LP count** for that market. When an LP in a given market reduces their commitment amount to 0 or gets closed out the LP count for that market gets decremented by 1.

> Once the active LP count in a market reaches the limit no further LP commitments are accepted. Once the count drops below the limit the LP commitments are accepted again.

Author: I think we will need to complicate this design a bit more to not allow a possibility of a market becoming deadlocked in a liquidity auction. 

So if market is in liquidity auction and the max LP limit is reached we should probably allow new LPs to kick out (their commitment gets cancelled) existing ones if their commitment is larger (in the commitment amount & seniority priority - we traverse the list by lowest commitment and within any commitment amount by time in market).

Reviewer 2: That sounds essential, yes. In fact why just liq. auction? I'd say that at any point in time we are sitting at the max, if a newly arriving LP would commit enough stake so that they have more ELS than the LP with the least ELS then the smallest LP is booted out.

Author: Added now.

### Change of network parameter

If the network parameter gets increased via a governance vote no further actions are needed.\
If it gets decreased below the current total number of LPs in the market then no further actions are needed. Specifically: all LPs active in the market remain unaffected, however no new liquidity provisions are accepted until the total LP count drops below the new limit.

## Acceptance Criteria

> * (<a name="0078-NWLI-001" href="#0078-NWLI-001">0078-NWLI-001</a>)

Reviewer 2: ;-)

Introduce a new network parameter `limits.markets.maxLimitOrders` controlling the maximum number of limit orders that can rest on a book in any given market. If the limit is reached limit orders can still be submitted, but:

* if the order is aggressive (results in a trade) the trade proceeds as normal,

> * if the order is passive (it doesn't match on entry) and would get added to the order book (it's a persistent order) it gets rejected.

Reviewer: Well, it would still be easy to submit very small order, which would match on entry and stay on the book just by using 2 different vega wallets right?

Author: I don't think so, I've clarified that part now, but not sure if that's what you were after.
